python/lsst/ctrl/bps/parsl/job.py

```diff
@@ -163,8 +163,17 @@ def evaluate_command_line(self, command: str) -> str:
             Command ready for execution on a worker.
         """
         command = command.format(**self.generic.cmdvals)  # BPS variables
-        command = re.sub(_env_regex, r"${\g<1>}", command)  # Environment variables
-        command = re.sub(_file_regex, lambda match: self.file_paths[match.group(1)], command)  # Files
+
+        # Make sure *all* symbolic names are resolved.
+        #
+        # In general, actual values for some symbolic names may contain other
+        # symbolic names. As a result, more than one iteration may be required
+        # to resolve all symbolic names.  For example, an actual value for
+        # a filename may contain a symbolic name for an environment variable.
+        while re.search(_env_regex, command) or re.search(_file_regex, command):
```

Comment:
Just a thought, but instead of repeating the search here and in the `re.sub` lines, maybe another option for the while is to loop until `command` stops changing? ie something like (untested):

```python
prev_command = command
while True:
    command = re.sub(...)
    command = re.sub(...)
    if command == prev_command:
        break
    prev_command = command
```
There may be a neater way to organize things but it seems better than running the regexes twice as often.

Reply:
Sure, I could use this version instead.

BTW, just out of curiosity, I timed both version with `timeit`. For a fairly short string like for example
```
<ENV:CTRL_MPEXEC_DIR>/bin/pipetask --long-log --log-level=VERBOSE pre-exec-init-qbb <FILE:butlerConfig> <FILE:runQgraphFile>
```
your version is ~10% faster than mine, i.e., it took 11.41 s to complete 1,000,000 executions for my version and 10.31 s  for yours.

Reply:
So most of the overhead is in the `re.sub`. I do think that "no string was replaced" is a neater end state since that's what you are actually wanting.